lib/mini_racer/truffleruby.rb

```diff
@@ -0,0 +1,371 @@
+
+      str = encode(str)
+      begin
+        translate do
+          eval_in_context(str, filename)
+        end
+      rescue Polyglot::ForeignException => e
+        raise RuntimeError, e.message, e.backtrace
+      rescue ::RuntimeError => e
+        if @current_exception
+          e = @current_exception
+          @current_exception = nil
+          raise e
+        else
+          raise e, e.message
+        end
+      rescue SystemStackError => e
+        raise RuntimeError, e.message
+      end
+    ensure
+      @entered = false
+    end
+
+    def call_unsafe(function_name, *arguments)
+      @entered = true
+      if !@has_entered && @snapshot
+        src = encode(@snapshot.instance_variable_get(:source))
+        begin
+          eval_in_context(src)
+        rescue Polyglot::ForeignException => e
+          raise RuntimeError, e.message, e.backtrace
+        end
+      end
+      @has_entered = true
+      raise RuntimeError, "TruffleRuby does not support call after stop" if @stopped
+      begin
+        translate do
+          function = eval_in_context(function_name)
+          function.call(*convert_ruby_to_js(arguments))
+        end
+      rescue Polyglot::ForeignException => e
+        raise RuntimeError, e.message, e.backtrace
+      rescue SystemStackError => e
+        raise RuntimeError, e.message
+      end
+    ensure
+      @entered = false
+    end
+
+    def create_isolate_value
+      # Returning a dummy object since TruffleRuby does not have a 1-1 concept with isolate.
+      # However, code and ASTs are shared between contexts.
+      Isolate.new
+    end
+
+    def isolate_mutex
+      @isolate_mutex
+    end
+
+    def translate
+      convert_result_to_ruby yield
+    rescue Object => e
+      message = e.message
+      if @current_exception
+        raise @current_exception
+      elsif e.message.start_with?('SyntaxError:')
+        error_class = MiniRacer::ParseError
+      elsif e.is_a?(MiniRacer::ScriptTerminatedError)
+        error_class = MiniRacer::ScriptTerminatedError
+      else
+        error_class = MiniRacer::RuntimeError
+      end
+
+      if error_class == MiniRacer::RuntimeError
+        bls = e.backtrace_locations&.select { |bl| bl.is_a?(Truffle::Interop::BacktraceLocation) && bl.source_location.language == 'js' }
+        if bls && !bls.empty?
+          if '(eval)' != bls[0].path
+            message = "#{e.message}\n at #{bls[0]}\n" + bls[1..].map(&:to_s).join("\n")
+          else
+            message = "#{e.message}\n" + bls.map(&:to_s).join("\n")
+          end
+        end
+        raise error_class, message
+      else
+        backtrace = e.backtrace.map { |line| line.sub('(eval)', '(mini_racer)') }
+        raise error_class, message, backtrace
+      end
+    end
+
+    def convert_result_to_ruby(value)
+      if is_object_or_array(value)
+        JSON.parse(js_json_stringify(value))
```

Comment:
Do we need this or can we just use `convert_js_to_ruby`?

Reply:
Updating to `convert_js_to_ruby` does not cause any test failures.

 I thought it might be related to security or something: https://github.com/rubyjs/mini_racer/blob/9cf4ed6d9290b3bcaf6e1cecf96dbf2a3166cc36/ext/mini_racer_extension/mini_racer_extension.cc#L425

I've left this as is for now. Let me know if I should change it.

Reply:
I think that's just a way to convert easily, but it's slow.
I prefer to only have one way to convert, so please remove the "through JSON" way then.